models/repo_list.go

```diff
@@ -167,18 +170,35 @@ func SearchRepositoryByName(opts *SearchRepoOptions) (RepositoryList, int64, err
 				accessCond = accessCond.Or(collaborateCond)
 			}
 
+			if opts.AllPublic {
+				accessCond = accessCond.Or(builder.Eq{"is_private": false})
+			}
+
 			cond = cond.And(accessCond)
 		}
 	}
 
-	if opts.OwnerID > 0 && opts.AllPublic {
-		cond = cond.Or(builder.Eq{"is_private": false})
-	}
-
 	if opts.Keyword != "" {
 		cond = cond.And(builder.Like{"lower_name", strings.ToLower(opts.Keyword)})
 	}
 
+	var forkCond builder.Cond
+	var mirrorCond builder.Cond
+
+	if opts.Fork != util.OptionalBoolNone {
+		forkCond = builder.Eq{"is_fork": opts.Fork == util.OptionalBoolTrue}
+	}
+
+	if opts.Mirror != util.OptionalBoolNone {
+		mirrorCond = builder.Eq{"is_mirror": opts.Mirror == util.OptionalBoolTrue}
+	}
+
+	if opts.Fork == util.OptionalBoolTrue && opts.Mirror == util.OptionalBoolTrue {
+		cond = cond.And(builder.Or(forkCond, mirrorCond))
```

Comment:
Why treat this as a special case? If `opts.Fork` and `opts.Mirror` are true, then caller is asking for repos that are both a fork AND a mirror, not repos that are a fork OR a mirror.

Reply:
I think repository can't be fork and mirror at the same time.

Reply:
I believe you are right that repos can't be a fork and a mirror. Nonetheless, if `opts.Fork` and `opts.Mirror` are true, then the caller is asking for repos that are both a fork AND a mirror, and the function should return what the user is asking for (which in this case will be the empty list).

The reason that I'm opposed to making a special case here is that (IMO) `opts.Fork` and `opts.Mirror` should be independent of each other. Setting `opts.Fork` to true should mean that the function will only return forks, regardless of what `opts.Mirror` is (and vice versa). Making special cases like "Only forks will be returned if `opts.Forks` is true, unless `opts.Mirror` is also true, in which case non-fork mirrors will also be returned" introduces unnecessary complexity and opportunities for bugs.

Reply:
I can change it, but with this you can do more. Using `and` operator has no viable use case, but if you use `or`, you can search for repositories that are `fork` or `mirror` or `fork || mirror` or `not fork && not mirror`. I don't see any opportunities for bugs here. You should know what you are doing If you ask for something. Same as you use `private` you are asking not only private repositories, options are not (or in some cases) exclusive.

Reply:
@ethantkoenig Changed as you wanted.